onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorNonZero.cpp

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "precomp.h"
+#include "core/providers/dml/DmlExecutionProvider/src/MLOperatorAuthorImpl.h"
+#include "core/providers/dml/DmlExecutionProvider/src/ExecutionProvider.h"
+
+namespace Dml
+{
+
+class DmlOperatorNonZero: public DmlOperator
+{
+public:
+    DmlOperatorNonZero(const MLOperatorKernelCreationContext& kernelCreationContext)
+    :   DmlOperator(kernelCreationContext)
+    {
+        ML_CHECK_VALID_ARGUMENT(kernelCreationContext.GetInputCount() == 1);
+        ML_CHECK_VALID_ARGUMENT(kernelCreationContext.GetOutputCount() == 1);
+
+        std::vector<DimensionType> inputShape = kernelCreationContext.GetTensorShapeDescription().GetInputTensorShape(0);
+
+        // Scalars have a rank of 0, but DML only supports 1 and more, which is the same
+        if (inputShape.empty())
+        {
+            inputShape.push_back(1);
+        }
+
+        uint32_t numElements = ComputeElementCountFromDimensions(inputShape);
+
+        gsl::span<const uint32_t> inputShapes[1] = {inputShape};
+        DmlOperator::InitializeWithShapes(kernelCreationContext, std::nullopt, std::nullopt, inputShapes, std::nullopt, 1);
+
+        m_rank = static_cast<DimensionType>(inputShape.size());
+        std::vector<DimensionType> outputCountShape = {1};
+        std::vector<DimensionType> outputCoordinatesShape = {numElements, m_rank};
+
+        // TODO: Remove the doubled strides when DML supports native int64 for NonZero
+        // TensorFlow outputs {rank, numElements}, but DML outputs {numElements, rank}
+        std::vector<DimensionType> outputCoordinatesStrides = {2, numElements * 2};
+        m_intermediateTensorDescs = {
+            TensorDesc(DML_TENSOR_DATA_TYPE_UINT32, outputCountShape),
+            TensorDesc(DML_TENSOR_DATA_TYPE_UINT32, outputCoordinatesShape, outputCoordinatesStrides),
+        };
+
+        // If the input has no elements, bypass the DML execution
```

Comment:
DML really should just treat these cases as a nop, rather than each framework needing to work around it. Plus, some workarounds in the DML EP break existing ONNX backend test cases like Padding with empty inputs which should yield a non-empty output. (but that's future work outside the scope of this PR...)